src/utils/viewer/ObjectPickingHelper.cpp

```diff
@@ -0,0 +1,132 @@
+  selectionFramebuffer_ = Mn::GL::Framebuffer{{{}, viewportSize}};
+  selectionFramebuffer_
+      .attachRenderbuffer(Mn::GL::Framebuffer::BufferAttachment::Depth,
+                          selectionDepth_)
+      .attachRenderbuffer(Mn::GL::Framebuffer::ColorAttachment{1},
+                          selectionDrawableId_);
+  return *this;
+}
+
+ObjectPickingHelper& ObjectPickingHelper::prepareToDraw() {
+  selectionFramebuffer_.bind();
+  mapForDraw();
+  selectionFramebuffer_.clearDepth(1.0f).clearColor(1, Mn::Vector4ui{0xffff});
+  CORRADE_INTERNAL_ASSERT(
+      selectionFramebuffer_.checkStatus(Mn::GL::FramebufferTarget::Draw) ==
+      Mn::GL::Framebuffer::Status::Complete);
+
+  // remove any visualized object that is picked before
+  if (meshVisualizerDrawable_) {
+    delete meshVisualizerDrawable_;
+    meshVisualizerDrawable_ = nullptr;
+  }
+
+  return *this;
+}
+
+ObjectPickingHelper& ObjectPickingHelper::mapForDraw() {
+  selectionFramebuffer_.mapForDraw({{Mn::Shaders::Generic3D::ColorOutput,
+                                     Mn::GL::Framebuffer::DrawAttachment::None},
+                                    {Mn::Shaders::Generic3D::ObjectIdOutput,
+                                     Mn::GL::Framebuffer::ColorAttachment{1}}});
+  return *this;
+}
+
+ObjectPickingHelper& ObjectPickingHelper::handleViewportChange(
+    Mn::Vector2i viewportSize) {
+  recreateFramebuffer(viewportSize);
+  selectionFramebuffer_.setViewport({{}, viewportSize});
+
+  shader_->setViewportSize(Mn::Vector2{viewportSize});
+  return *this;
+}
+
+unsigned int ObjectPickingHelper::getObjectId(
+    Magnum::Vector2i mouseEventPosition,
+    Magnum::Vector2i windowSize) {
+  selectionFramebuffer_.mapForRead(Mn::GL::Framebuffer::ColorAttachment{1});
+  CORRADE_INTERNAL_ASSERT(
+      selectionFramebuffer_.checkStatus(Mn::GL::FramebufferTarget::Read) ==
+      Mn::GL::Framebuffer::Status::Complete);
+
+  // First scale the position from being relative to window size to being
+  // relative to framebuffer size as those two can be different on HiDPI
+  // systems
+  const Mn::Vector2i position =
+      mouseEventPosition *
+      Mn::Vector2{selectionFramebuffer_.viewport().size()} /
+      Mn::Vector2{windowSize};
+
+  const Mn::Vector2i fbPosition{
+      position.x(),
+      selectionFramebuffer_.viewport().sizeY() - position.y() - 1};
+  const Mn::Range2Di area = Mn::Range2Di::fromSize(fbPosition, Mn::Vector2i{1});
+
+  const Mn::UnsignedInt pickedObject =
+      selectionFramebuffer_.read(area, {Mn::PixelFormat::R32UI})
+          .pixels<Mn::UnsignedInt>()[0][0];
+
+  return pickedObject;
+}
+
+ObjectPickingHelper& ObjectPickingHelper::createPickedObjectVisualizer(
+    esp::gfx::Drawable* pickedObject) {
+  if (meshVisualizerDrawable_) {
+    delete meshVisualizerDrawable_;
+    meshVisualizerDrawable_ = nullptr;
+  }
+
+  if (!pickedObject) {
+    return *this;
+  }
+
+  // magnum scene graph will handle the recycling even we did not recycle it by
```

Comment:
> recycling

you mean ... "garbage collection"? :) i never saw this term used in this context

Reply:
Yes, garbage collection. The scene graph will destroy all the features attaching a node. 
That is what I meant.